### PR TITLE
Improve build process

### PR DESCRIPTION
**Pull request by frarees**
Quite some internal Makefile rework, but in the end this brings manual libpng searching via `PNG` (in case you want to avoid pkg-config), exposes some more variables on the Makefile and updates the README to reflect all those changes.

It explains build intructions with bigger detail, and warns about yacc as a requirement for rgbasm.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,14 +1,31 @@
-PKG_CONFIG =	pkg-config
-WARNFLAGS =	-Wall -Werror=implicit
-PNGFLAGS !=	${PKG_CONFIG} --cflags libpng
-REALCFLAGS =	${CFLAGS} ${WARNFLAGS} ${PNGFLAGS} -Iinclude -g \
-		-std=c99 -D_POSIX_C_SOURCE=200809L
+CC      = gcc -std=c99 -D_POSIX_C_SOURCE=200809L
```
Comment (bentley) [inline Makefile]: Please don’t hardcode gcc here.

File: Makefile [hunk 2]
```diff
@@ -1,14 +1,31 @@
...
+YACC    = yacc
```
Comment (bentley) [inline Makefile]: This is already implicit (see POSIX).

File: Makefile [hunk 3]
```diff
@@ -1,14 +1,31 @@
...
-Q =		@
+PKGCONFIG = pkg-config
+PREFIX    = /usr/local
+BINPREFIX = ${PREFIX}/bin
+MANPREFIX = ${PREFIX}/man
+Q         = @
+
+# libraries
+
+ifdef PNG
```
Comment (bentley) [inline Makefile]: `ifdef` is not portable make.

File: Makefile [hunk 4]
```diff
@@ -1,14 +1,31 @@
...
+	PNG_LDFLAGS = -L${PNG}/lib
+	PNG_LIBS    = -lpng
+else
+	# requires pkg-config
+	PNG_CFLAGS  = $(shell ${PKGCONFIG} --cflags libpng)
+	PNG_LDFLAGS = $(shell ${PKGCONFIG} --libs-only-L libpng)
+	PNG_LIBS    = $(shell ${PKGCONFIG} --libs-only-l libpng)
+endif
+
+M_LIBS = -lm
```
Comment (bentley) [inline Makefile]: What’s a situation where a user would want to override libm?

File: Makefile [hunk 5]
```diff
@@ -1,14 +1,31 @@
...
+
+ifdef PNG
+	PNG_CFLAGS  = -I${PNG}/include
+	PNG_LDFLAGS = -L${PNG}/lib
+	PNG_LIBS    = -lpng
+else
+	# requires pkg-config
+	PNG_CFLAGS  = $(shell ${PKGCONFIG} --cflags libpng)
+	PNG_LDFLAGS = $(shell ${PKGCONFIG} --libs-only-L libpng)
+	PNG_LIBS    = $(shell ${PKGCONFIG} --libs-only-l libpng)
```
Comment (bentley) [inline Makefile]: `$(shell foo)` is not portable make.

File: README [hunk 6]
```diff
@@ -17,51 +17,68 @@ more like other UNIX tools.
...
 
-  make
+  brew install rgbds --build-from-source
 
-Then to install the compiled programs and manual pages, run (with appropriate
-privileges):
+or
 
-  make install
+  brew install rgbds --HEAD
```
Comment (bentley) [inline README]: Can you make this change in a separate pull request?

**Comment by bentley** [comment]
The PNG bits should already be fixed on OS X with #109. That said I’m not opposed to some build system rework to override it, but this pull request isn’t mergeable as‐is.

A simple configure script might even be appropriate, but autotools is way too heavy a dependency for my tastes.

**Comment by frarees** [comment]
Thanks for the suggestions.

I am quite unaware of portable makefiles and where to find proper documentation. Even books on the subject rely on conditionals (ifdef) and so on. So, I've tried to let it be more flexible and less _automatic_ (check README for _updated_ usage)

I've reworked the Makefile, hope I am doing it in the right direction. Let me know what do you think.

The README has been updated too, and if this PR leads nowhere, I will open another one specifically about it.

About autotools, you're right it can feel heavy for such a simple compilation process. However, if you're ever interested, check out my `autotools` branch an implementation.

**Comment by AntonioND** [comment]
I'm not sure about just removing the libpng flags. Maybe it's worth having a makefile for each main OS.

**Comment by Ben10do** [comment]
That really doesn't sound like a good idea to me – it'd be far easier to maintain a single portable makefile.

BTW, the Readme should be updated to fit the new Markdown format, whether it's in this PR or another.

**Comment by AntonioND** [comment]
Yeah, but then removing the libpng flags will be quite annoying for Linux users. Right now I just have to type `make` and everything just works. Anything else would be a step back I think.

**Comment by frarees** [comment]
What do you think we should do with this?

Should we just add some variables for libs and let users go in and edit them for their specific case?

I'd just want to point out that it is interesting that we retain somehow the ability to specify those from outside, so that applications like macports or homebrew (in case of macOS) can provide their own versions.

**Comment by AntonioND** [comment]
As far as I know you can specify the value of variables from the command line, so that shouldn't be a problem.

I think that the only thing we really need is to remove the `WARNFLAGS`, `REALCFLAGS` and things like that.

**Comment by AntonioND** [comment]
This PR is quite outdated so I'm going to close it. A similar PR is in regards to the Makefile is here https://github.com/rednex/rgbds/pull/176. The changes in the README are quite outdated as well.
